python/pyspark/daemon.py

```diff
@@ -102,7 +102,7 @@ def shutdown(code):
         signal.signal(SIGTERM, SIG_DFL)
         # Send SIGHUP to notify workers of shutdown
         os.kill(0, SIGHUP)
```

Comment:
Since many other lines in this module are already using `os._exit()`, it is probably fine to use it here too.  I do wonder, though, if inserting a slight pause between the above call to `os.kill(0, SIGHUP)` and the call to exit might relieve the underlying issue, and allow proper cleanup in the children.

Reply:
They are different because here it should guarantee to let their child die (forked workers). Other two occurrences of it is just to the single forked worker to die so cleanup procedure is not critical there.

`os._exit()` call itself does not guarantee those cleanups as pointed out and [as documented](https://docs.python.org/3/library/os.html#os._exit).

Reply:
@HyukjinKwon , so do you think a pause, such as `time.sleep(3)` would help?  Or even doing the effort of a time-limited `wait` for all children? (It should not wait indefinitely in this situation, I think.)

Reply:
From a cursory look, it might help but I don't know if it will virtually help or not since we don't exactly know what's the delta between Python 3.x and Python 3.7 here. I would like to avoid to use `os._exit()` which is officially discouraged in this way.

If we should, I would like to know if it was really a bug in Python 3.7.